netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package io.grpc.netty;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.ReferenceCountUtil;
+
+public class WriteCombiningHandler extends ChannelOutboundHandlerAdapter {
+
+  private static final int BUFFER_SIZE = 4096;
+
+  private ByteBuf buffer;
+  private ByteBufAllocator allocator;
+
+  @Override
+  public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+    allocator = ctx.channel().alloc();
+  }
+
+  @Override
+  public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+    if (buffer != null) {
+      ReferenceCountUtil.safeRelease(buffer);
+      buffer = null;
+    }
+  }
+
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+    ByteBuf data = (ByteBuf) msg;
+    if (buffer == null) {
+      buffer = allocator.directBuffer(BUFFER_SIZE, BUFFER_SIZE);
+    }
+
+    if (buffer.writableBytes() >= data.readableBytes()) {
+      buffer.writeBytes(data);
+      promise.setSuccess();
```

Comment:
@nmittler I saw that, but this situation is a bit different. 1) We need to maintain the actual promises and not just a count 2) We need to complete the promises as their respective buffers are written to the network. That is, we need a mapping of byte ranges to promises (promise 1 completed once bytes [0,100] have been written, promise 2 once bytes [101, 150] have been written and so on.). Now we sure don't need to complete each promise immediately once its corresponding bytes have been flushed, but we still need to handle the case where only a fraction of the aggregated ByteBuf has been flushed. Good thing Netty already comes with this capability, namely [ChannelProgressivePromise](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L235). So I think this makes this quite straightfoward.